### PR TITLE
Creating volume attachments

### LINKED ISSUE #309: Add volume_attachment property to Server Profiles and SP Templates so VAs can be more easily managed
### Scenario/Intent
As mentioned by @aalexmonteiro in PR #308 :  
*In the Ruby SDK there is a method for creating a server profile with a volume attachment add_volume_attachment. Would it be nice to have a volume_attachment property?*

It'd likely be a good improvement if we can use some method to simply creation and management of VAs in Chef, if that is possible.

### Expected Result
Have a simpler/better way to manage volume attachments in Chef

### Actual Result
Currently we're dependant on the actual REST API formatting

### PR DESCRIPTION
**Pull request by ricardogpsf**
### Description
Added helper method to be used by Server Profile and Server Profile Template to attach Volumes.
The method can create a Volume and attach it or just attach a already created Volume, just based on the information of `volume_attachments` property of oneview_server_profile resource or oneview_server_profile_template resource.

The tests about volume attachments are shared by oneview_server_profile and oneview_server_profile_template.

### Issues Resolved
#309 

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass (`$ rake test`).
  - [ ] New resources and/or actions have are included in the [matchers.rb](https://github.com/HewlettPackard/oneview-chef/blob/master/libraries/matchers.rb) and [matchers_spec](https://github.com/HewlettPackard/oneview-chef/blob/master/spec/unit/resources/matchers_spec.rb).
- [x] New functionality has been documented in the README if applicable.
  - [x] New functionality has been thoroughly documented in [examples](https://github.com/HewlettPackard/oneview-chef/tree/master/examples/) (please include helpful comments).
- [x] Changes documented in the [CHANGELOG](https://github.com/HewlettPackard/oneview-chef/blob/master/CHANGELOG.md).

### CODE REVIEW

File: README.md
Comment (fgbulsoni) [inline README.md]: ?

File: README.md
Comment (fgbulsoni) [inline README.md]: Likely rewrite, it is kind of confusing. Be clear and concise.

File: README.md
Comment (fgbulsoni) [inline README.md]: key-pair data to be the specific attributes of **the** Oneview Volume Attachment

File: examples/server_profile.rb
Comment (fgbulsoni) [inline examples/server_profile.rb]: commented out vars?

**Review by fgbulsoni — CHANGES_REQUESTED** [review]

File: examples/server_profile.rb
Comment (fgbulsoni) [inline examples/server_profile.rb]: I feel like this example is becoming heavily poluted by volume attachments. Not sure if that is just my impression? 
This first part became all about VA variables. Perhaps a separated example is in order for this?

Comment (fgbulsoni) [inline examples/server_profile.rb]: Perhaps these data samples could be loaded from a different file, and we could have a comment here for instructing the user on that?
It feels like a lot of visual noise from VAs data when I want to see a server profile example.

File: examples/server_profile_template.rb
Comment (fgbulsoni) [inline examples/server_profile_template.rb]: Same here, lots of visual noise from VA.

File: README.md
Comment (ricardogpsf) [reply README.md]: I forgot this "noise". =(

File: examples/server_profile.rb
Comment (ricardogpsf) [reply examples/server_profile.rb]: 😦 😅  I'll change it

File: libraries/resource_providers/api200/server_profile_provider.rb
Comment (fgbulsoni) [inline libraries/resource_providers/api200/server_profile_provider.rb]: Probably rewrite message.

Comment (ricardogpsf) [reply libraries/resource_providers/api200/server_profile_provider.rb]: Yeah, maybe ~`"To add volume attachments you need inform volume or volume_data"`~ `"To add volume attachments you need to specify'volume' or 'volume_data' insides 'volume_attachments' options"` is good.

File: libraries/resource_providers/api200/server_profile_provider.rb
Comment (fgbulsoni) [inline libraries/resource_providers/api200/server_profile_provider.rb]: (...) with **an** attachment you need **to specify** the 'storage_system' and 'storage_pool' names inside 'options'.

File: README.md
Comment (ricardogpsf) [reply README.md]: What about "To attach a Volume already created, put into the 'volume_attachments' something like:" ?

File: examples/server_profile.rb
Comment (ricardogpsf) [reply examples/server_profile.rb]: I'm thinking move all examples of volume attachments to another file, not just the data samples, what do you think?

Comment (ricardogpsf) [reply examples/server_profile.rb]: Thinking better, I'll just put the variables within separate file, because looks more simple, and it avoids create similar files, with similar examples and code of variables duplicated.

File: attributes/volume_attachments_variables.rb
Comment (fgbulsoni) [inline attributes/volume_attachments_variables.rb]: commented out?

File: attributes/volume_attachments_variables.rb
Comment (fgbulsoni) [inline attributes/volume_attachments_variables.rb]: commented out?

File: attributes/volume_attachments_variables.rb
Comment (ricardogpsf) [reply attributes/volume_attachments_variables.rb]: @fgbulsoni to make it easier for clients to run the example with api500, just by uncommenting that lines.

File: libraries/resource_providers/api200/server_profile_provider.rb
Comment (fgbulsoni) [reply libraries/resource_providers/api200/server_profile_provider.rb]: `A 'volume' or 'volume_data' is required for each volume attachment specified.`

File: examples/server_profile.rb
Comment (fgbulsoni) [inline examples/server_profile.rb]: Probably better to be specific here:

This example requires the following resources to be available in the appliance:
- FC Networks: 'FC1', 'FC2'
- Storage Pool: (...)

Comment (ricardogpsf) [reply examples/server_profile.rb]: @fgbulsoni Maybe the below text is good, it is simple to understand?
```ruby
# NOTES:
# This example requires the following resources to be available in the appliance:
#  - FC Network: 'FCNetwork1'
#  - FCoE Network: 'fcoe1'
#  - Ethernet Network: 'EthernetNetwork1'
#  - Server Profile Template: 'spt1'
#  - Server Hardware Type: 'BL660c Gen9 1'
#  - Server Hardware: 'Encl1, bay 2'
#  - Enclosure Group: 'EG_1'
# To create volume attachments:
#  - The attributes file "volume_attachments_variables" is loading variables to be used in this example.
#  - The Storage System must have at least one connection using 'FCNetwork1' and this same network network must have uplinkSet connected on the Interconnect,
#   and the Server Hardware related to that network network is the Server Hardware used in this example.
```

Comment (fgbulsoni) [reply examples/server_profile.rb]: (...) this same *network network* (...)
must have **an** uplinkSet (...)
(...)Server Hardware related to that *network network*
Also I'd suggest using names following a pattern. i.e. stick to either 
- `FCNetwork1`, `FCoENetwork1` 
or 
- `fc1`, `fcoe1`
, etc. Not both styles.

**Review by fgbulsoni — APPROVED** [review]
LGTM, but since this PR contains many modifications I'd get @tmiotto 's approval as well before merging.

**Comment by ricardogpsf** [comment]
ok!
@tmiotto can you review, please?

File: README.md
Comment (tmiotto) [inline README.md]: "when creating or updating"

File: README.md
Comment (tmiotto) [inline README.md]: same here

File: README.md
Comment (tmiotto) [inline README.md]: Is this intended to be empty?

If not, but you want to ommit the data use something like `attachment_data: { ... }`

File: attributes/volume_attachments_variables.rb
Comment (tmiotto) [inline attributes/volume_attachments_variables.rb:35]: I don't like how the examples were split.

Put the data in the same file as the example so the user won't need to alternate between multiple files to see on example.

If needed please create a separate `volume_attachment.rb` example, and if needed create one for API500.

File: libraries/resource_providers/api200/server_profile_provider.rb
Comment (tmiotto) [inline libraries/resource_providers/api200/server_profile_provider.rb]: inside*

File: libraries/resource_providers/api200/server_profile_provider.rb
Comment (tmiotto) [inline libraries/resource_providers/api200/server_profile_provider.rb]: This clause is useless since you are defaulting the volume_attachments to `[]`, therefore it would never return here.

File: spec/unit/resources/server_profile/create_spec.rb
Comment (tmiotto) [inline spec/unit/resources/server_profile/create_spec.rb]: ` OneviewSDK::API200::ServerProfile`

File: spec/unit/shared_examples/volume_attachments.rb
Comment (tmiotto) [inline spec/unit/shared_examples/volume_attachments.rb]: If you are using a matcher just match the relevant part of the error message, this way you won't need to change the test if the message in the code is changed (That is the case here)

**Review by tmiotto — CHANGES_REQUESTED** [review]

File: libraries/resource_providers/api200/server_profile_provider.rb
Comment (ricardogpsf) [reply libraries/resource_providers/api200/server_profile_provider.rb]: Even if user set the value to nil?

Comment (ricardogpsf) [reply libraries/resource_providers/api200/server_profile_provider.rb]: really it is not necessary, if user is setting the value to nil, he is using the recipe in wrong way.

Comment (tmiotto) [reply libraries/resource_providers/api200/server_profile_provider.rb]: By one side the user really wants to set it to nil regardless of the recommendations to leave it alone when he does not wants any attachment, thus becoming susceptible to unexpected failures. But, by other side, the user may inadvertedly put it that way, so feel free to change it or not.

File: attributes/volume_attachments_variables.rb
Comment (ricardogpsf) [reply attributes/volume_attachments_variables.rb:35]: I'll split the examples in 4:
- server_profile
- server_profile_template
- server_profile_with_attachments
- server_profile_template_with_attachments

I think that is simples and organized. What do you think?

Comment (tmiotto) [reply attributes/volume_attachments_variables.rb:35]: You can have just one example for server_profile_with_attachments and server_profile_template_with_attachments since they'll basically be the same.

So it'll leave you with 3 files:
- server_profile
- server_profile_template
- server_profiles_with_attachments (Or you can simply put these examples inside the previous files, leaving you with two example files)

And, if needed add one more for API500 if the data or behavior differ from the previous ones.

Comment (ricardogpsf) [reply attributes/volume_attachments_variables.rb:35]: ok, the original solution was keeping the volume attachments examples inside the previous files of examples, but the data necessary to run the exemple of attachments made much noises into the files, so we decided move the data to other file. But I think that the solution with 3 files seems good, because the third file will have just eamples about volume attachments, maybe can be named as volume_attachments.rb.
What do you think?

**Review by tmiotto — APPROVED** [review]

**Pull request merged**
